recipes/flatcc/all/test_package/conanfile.py

```diff
@@ -0,0 +1,18 @@
+import os.path
+
+from conans import ConanFile, CMake, tools
+
+
+class FlatccTestConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
```

Comment:
If the reason you're adding the environment variables `LD_LIBRARY_PATH`/`DYLD_LIBRARY_PATH` is to let `cmake.configure` work, then you can add these by using a RunEnvironment.
```suggestion
        with tools.environment_append(RunEnvironment(self).vars):
            cmake = CMake(self)
            cmake.configure()
            cmake.build()
```

Reply:
Yes, that is indeed the reason.

For the test package it is no problem to use a RunEnvironment. But now all the packages that use flatcc also have to know that they should run the cmake commands using a RunEnvironment. That makes things more complicated than necessary IMHO.
Is it really a problem to add the `LD_LIBRARY_PATH/DYLD_LIBRARY_PATH` variables in package_info() ?

Reply:
@jgsogo it would be useful, I think, to add a `run_environment` option to all build helpers.
It would be useful for at least autotools, cmake and Medion.

Reply:
@wdobbe I'm not sure this would work on macos, due to System Integration Protection.

See https://github.com/conan-io/conan-center-index/pull/1179#issuecomment-621339628 (and the comments around) for my personal fight against SIP

Reply:
Hi! The problem with Macos and SIP is not solvable right now. Even if those variables are in the environment while running CMake, they won't be accessible inside or to those _custom commands_ (check it with a shared build if `flatcc` has dependencies, or check the issue linked by @madebr ). 

An option would be for the build helpers to populate the environment with their build-requires, I agree, but it won't solve the issue with Mac and SIP 😞 

---

I have a pending POC for this: the problem is not only Mac and SIP, but Conan populating the environment with **all the variables from all the build-requirements**, this is a problem because some variables may override values from other build-requires,... but also, if build-requires are independent graphs, some libraries could appear with different versions in different build-requires, and the system will find the first one according to the paths in `DY/LD_LIBRARY_PATH` 😱 

Alternatives? Executables are listed in the `cpp_info` attribute (https://github.com/conan-io/conan/issues/7240), then Conan creates a _wrapper_ for every executable, something like (for `protobuf` build-requires):

**protoc.bat**
```
activate environment
path/to/build-requires/executable/protoc -- forward all arguments
deactivate environment
```

And these wrappers are added to the `PATH`, Macos and SIP doesn't clean the `PATH` variable, so a custom command `protoc ...` will find and execute that wrapper, environment from different build-requires is not mixed, everything clean, everything works out-of-the-box.

wdyt? Do you think it would be a nice approach? Will it work?

---

Now, answering the question here, `flatcc` will be used mainly as a build-requires, Conan already populates the environment with the values of `DY/LD....` taken from the `cpp_info.bin` attribute when using two profiles (new approach with the build and host contexts). So there is no need to add these variables to `cpp_info.env_info`, only add the `bin` directory to the `PATH`

Reply:
Even when I add "build_requires = "flatcc/[*]" to the test package, the flatcc doesn't find its shared libraries without the LD_LIBRARY_PATH.

Moreover, the flatcc is not only a build requirement but also a runtime requirement because the flatcc package creates library flatccrt that generated code needs to link to.

So for now I will leave the (DY)LD_LIBRARY_PATH lines in package_info() unless someone knows a working solution that doesn't require those lines.

Reply:
@jgsogo 
Are you sure that such a `protoc.bat`/`protoc.sh` will work on Macos? The `activate environment` will also use `DYLD_LIBRARY_PATH`.


---

How about fixing this problem on cmake as follows?

Let the cmake helpers create a cmake variable named e.g. `CONAN_LIBRARY_PATHS` that appends all `*_LIBRARY_PATHS`.
On `Macos` it would contain: `DYLD_LIBRARY_PATH=/path/to/lib1:/path/to/lib2/:...`

Then a user can do:
```
add_custom_command(... COMMAND ${CONAN_LIBRARY_PATHS} program)
```


> Now, answering the question here, flatcc will be used mainly as a build-requires, Conan already populates the environment with the values of DY/LD.... taken from the cpp_info.bin attribute when using two profiles (new approach with the build and host contexts). So there is no need to add these variables to cpp_info.env_info, only add the bin directory to the PATH

This would be really great! Imho, conan should have knowledge of a package's capability to be used as a requirement/build requirement.

Don't forget that some recipes also export environment variables to make it it work/override settings...

Reply:
Being a build-requirement and a requirement is not a problem. It will work the same way as `protobuf` does (it provides the library and the `protoc` executable). Populating `(DY)LD_LIBRARY_PATH` won't help if you call `flatcc` from inside CMake, those values will be cleaned before running the executable `flatcc` 🤷  And, IMHO, I think Macos is doing it right.

The approach with the wrappers work (at least in Macos, I assume Linux will work too). You can have a look at some brute-force implementation here (https://github.com/conan-io/conan/compare/develop...jgsogo:poc/xbuild-exec-wrapper) that works together with this example here (https://github.com/jgsogo/conan-xbuild-scenarios). I need to share these ideas with the team and evaluate if they make sense or not, but it works 😆  ...testing in Windows still pending.

I've also considered the CMake approach, but my main concern is that it would be exclusive to CMake and every other generator will need its own implementation. Not only the build systems ones, but also the virtualenvironments we can generate from Conan. With this _wrapper_ approach I only need to create these wrappers, prepend them to the `PATH` and it works out-of-the-box.

Reply:
Hi Javier,
Thanks for your work on the wrappers. I will leave Macos shared as an invalid configuration until the wrappers (or other solution) is available in Conan.

> Populating `(DY)LD_LIBRARY_PATH` won't help if you call `flatcc` from inside CMake, those values will be cleaned before running the executable `flatcc`

In the test_package flatcc is called from cmake and this works fine when LD_LIBRARY_PATH is set by the flatcc recipe, also in our production code that uses flatcc with this package.

Reply:
I think it is because `flatcc` doesn't have dependencies (or they are linked statically). The problem is when you need `DYLD_LIBRARY_PATH` to locate other `dylib` libraries.

Reply:
I was refering to Linux. flatcc called from cmake works fine even when flatcc is build shared.
My Mac is a hackintosh, looks like SIP is always off there, as I have no DYLD_LIBRARY_PATH problems. But unfortunately that also means I cannot test with SIP on.